### PR TITLE
Add FlatPageTableLib

### PR DESCRIPTION
**Pull request by TaylorBeebe**
## Description

Creates a new library, FlatPageTableLib, which works on X64 and AARCH64 platforms and converts the page table to a "flat" version. The flat version is a one-dimensional array where each entry is an address, a length, and attributes. The library will walk the page/translation table and combine blocks/leaves with the same attributes into a single entry in the flat array. The attributes mask for each architecture is defined in the header and includes both the upper and lower block/leaf attributes. On both X64 and AARCH64, the hierarchical inheritance of attributes is factored into the determination of block/leaf attributes. This allows the consumer of the library to easily check the attributes of any region in the page/translation table.

A future update will add "map" functionality which will allow the user to manipulate the attributes of the page/translation table with correct hierarchical updates. That functionality will require allocation to work because spitting blocks larger than 4K will require additional table memory.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Tested by running the paging audit on SBSA and Q35 and by comparing the output against the Memory Attribute Protocol

## Integration Instructions

N/A

### CODE REVIEW

File: MdeModulePkg/Library/FlatPageTableLib/X64/FlatPageTableX64.c [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Map->ArchSignature = X64_PAGE_MAP_SIGNATURE;
+
+  // Poll CR4 to deterimine the page table depth
+  Cr4.UintN = AsmReadCr4 ();
+
+  if (Cr4.Bits.LA57 != 0) {
+    PagingMode = Paging5Level;
```
Comment (VivianNK) [inline MdeModulePkg/Library/FlatPageTableLib/X64/FlatPageTableX64.c:51]: Where do these Paging5/4Level values come from?

File: MdeModulePkg/Include/Library/FlatPageTableLib.h [hunk 2]
```diff
@@ -0,0 +1,166 @@
+/** @file
```
Comment (VivianNK) [inline MdeModulePkg/Include/Library/FlatPageTableLib.h:1]: nit: Honestly not sure if we follow this but for the comment documentation tags @file, they should be followed by the filename.
e.g.
@file FlatPageTableLib.h

Or do we use a different system?

File: MdeModulePkg/Library/FlatPageTableLib/X64/FlatPageTableX64.c (hunk 1 above)
Comment (TaylorBeebe) [reply MdeModulePkg/Library/FlatPageTableLib/X64/FlatPageTableX64.c:51]: The definitions come from CpuPageTableLib

File: MdeModulePkg/Include/Library/FlatPageTableLib.h (hunk 2 above)
Comment (TaylorBeebe) [reply MdeModulePkg/Include/Library/FlatPageTableLib.h:1]: In EDK2 headers, I only see @file. In some Project Mu headers, I see @file <filename>. I'll have to check if the EDK2 code format guidelines weigh in on this :)

**Comment by TaylorBeebe** [comment]
Moved to mu_plus: https://github.com/microsoft/mu_plus/pull/322

**Pull request closed without merging**